C error: Generic map checking key exists.

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->

<details>
<summary>V version: V 0.4.11 6da1923, press to see full `v doctor` output</summary>

|V full version      |V 0.4.11 7dc3889f19e7bdc6fb803ad4d1d65fd9d50c6e0c.6da1923
|:-------------------|:-------------------
|OS                  |linux, Ubuntu 22.04.5 LTS
|Processor           |4 cpus, 64bit, little endian, Intel(R) Core(TM) i3-4150 CPU @ 3.50GHz
|Memory              |0.47GB/3.7GB
|                    |
|V executable        |/home/jorge/v/v
|V last modified time|2025-07-31 13:38:08
|                    |
|V home dir          |OK, value: /home/jorge/v
|VMODULES            |OK, value: /home/jorge/.vmodules
|VTMP                |OK, value: /tmp/v_1000
|Current working dir |OK, value: /home/jorge/bugs
|                    |
|Git version         |git version 2.34.1
|V git status        |weekly.2025.16-527-g6da19238
|.git/config present |true
|                    |
|cc version          |cc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
|gcc version         |gcc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
|clang version       |Ubuntu clang version 14.0.0-1ubuntu1.1
|tcc version         |tcc version 0.9.28rc 2025-02-13 HEAD@f8bd136d (x86_64 Linux)
|tcc git status      |thirdparty-linux-amd64 696c1d84
|emcc version        |emcc (Emscripten gcc/clang-like replacement + linker emulating GNU ld) 3.1.5 ()
|glibc version       |ldd (Ubuntu GLIBC 2.35-0ubuntu3.10) 2.35

</details>

**What did you do?**
`./v -g -o vdbg cmd/v && ./vdbg  250731.v && 250731`
```v
pub struct Set[T] {
mut:
        elements map[T]u8
}

pub fn (l Set[T]) == (r Set[T]) bool {
        if l.elements.len != r.elements.len {
                return false
        }
        for key, l_val in l.elements {
                r_val := r.elements[key] or {
                        return false
                }
                if l_val != r_val {
                        return false
                }
        }
        return true
}

fn main() {
	_ := Set[int]{}
}
```

**What did you see?**
```
================== C compilation error (from tcc): ==============
cc: /tmp/v_1000/250731.01K1GB9892BM7Q8K58Z4RPQCSE.tmp.c:6122: warning: implicit declaration of function '_v_error'
cc: /tmp/v_1000/250731.01K1GB9892BM7Q8K58Z4RPQCSE.tmp.c:6122: error: cannot convert 'int' to 'struct IError'
=================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
builder error: 
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```

**What did you expect to see?**

compiled program

Maybe related with https://github.com/vlang/v/issues/24983
> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/markused/walker.v
+++ b/vlib/v/markused/walker.v
@@ -628,7 +628,7 @@ fn (mut w Walker) expr(node_ ast.Expr) {
 			w.expr(node.expr)
 			w.mark_by_type(node.expr_type)
 			w.uses_guard = true
-			if node.expr_type != 0 && !w.uses_str_index_check
+			if node.expr_type != 0 && !w.uses_str_index_check && node.expr is ast.IndexExpr
 				&& w.table.final_sym(node.expr_type).kind in [.u8, .string] {
 				w.uses_str_index_check = true
 			}
@@ -1296,9 +1296,6 @@ fn (mut w Walker) mark_resource_dependencies() {
 	if w.uses_mem_align {
 		w.fn_by_name('memdup_align')
 	}
-	if w.uses_guard || w.uses_check_index {
-		w.fn_by_name('error')
-	}
 	if w.uses_spawn {
 		w.fn_by_name('malloc')
 		w.fn_by_name('tos3')
@@ -1406,6 +1403,9 @@ fn (mut w Walker) mark_resource_dependencies() {
 			w.fn_by_name(k)
 			continue
 		}
+	}
+	if w.uses_guard || w.uses_check_index {
+		w.fn_by_name('error')
 	}
 	if w.uses_append {
 		ref_array_idx_str := int(ast.array_type.ref()).str()

--- a/vlib/v/tests/generics/struct_generic_map_index_test.v
+++ b/vlib/v/tests/generics/struct_generic_map_index_test.v
@@ -1,0 +1,21 @@
+pub struct Set[T] {
+mut:
+	elements map[T]u8
+}
+
+pub fn (l Set[T]) == (r Set[T]) bool {
+	if l.elements.len != r.elements.len {
+		return false
+	}
+	for key, l_val in l.elements {
+		r_val := r.elements[key] or { return false }
+		if l_val != r_val {
+			return false
+		}
+	}
+	return true
+}
+
+fn test_main() {
+	_ := Set[int]{}
+}

